flask/app.py

```diff
@@ -1791,32 +1791,39 @@ def handle_url_build_error(self, error, endpoint, values):
         raise error
 
     def preprocess_request(self):
-        """Called before the actual request dispatching and will
-        call each :meth:`before_request` decorated function, passing no
-        arguments.
-        If any of these functions returns a value, it's handled as
-        if it was the return value from the view and further
-        request handling is stopped.
+        """Called before the request dispatching.
+        Triggers two set of hook functions that should be invoked prior to a request dispatching:
+        self.url_value_preprocessors and self.before_request_funcs.
+        In both cases, triggers only the functions that are either global or registered to the current blueprint.
 
-        This also triggers the :meth:`url_value_processor` functions before
-        the actual :meth:`before_request` functions are called.
+        If any of the before_request_funcs returns a value, it's handled as if it was
+        the return value from the view function, and further request handling is stopped.
         """
-        bp = _request_ctx_stack.top.request.blueprint
-
-        funcs = self.url_value_preprocessors.get(None, ())
-        if bp is not None and bp in self.url_value_preprocessors:
-            funcs = chain(funcs, self.url_value_preprocessors[bp])
```

Comment:
@davidism The duplication I was referring to isn't the for loops, but the three lines of code that come before each one.
They are exactly identical except for the variables, so IMO it makes sense to factor them out to a function.